hack/generator/pkg/astmodel/package_reference.go

```diff
@@ -36,3 +44,219 @@ func PackageAsLocalPackage(pkg PackageReference) (LocalPackageReference, error)
 	}
 	return LocalPackageReference{}, errors.Errorf("%q is not a local package", pkg.PackagePath())
 }
+
+func SortPackageReferencesByPathAndVersion(packages []PackageReference) {
+	sort.Slice(packages, func(i, j int) bool {
+		comparer := versionComparer{
+			left:  []rune(packages[i].PackagePath()),
+			right: []rune(packages[j].PackagePath()),
+		}
+		compare := comparer.Compare()
+		return compare < 0
+	})
+}
+
+// versionComparer captures our state while doing an alphanumeric version comparision
+// We need separate indexes for each side because we're doing a numeric comparison, which will
+// compare "100" and "0100" as equal (leading zeros are not significant)
+type versionComparer struct {
+	left       []rune
+	leftIndex  int
+	right      []rune
+	rightIndex int
+}
+
+func (v *versionComparer) Compare() int {
+	for {
+		if v.leftIndex >= len(v.left) && v.rightIndex >= len(v.right) {
+			// Ran out of both arrays at the same time
+			break
+		}
+
+		if v.leftIndex >= len(v.left) {
+			// Ran out of the left array; if the right array has an identifier, right goes first, otherwise left
+			if unicode.IsLetter(v.right[v.rightIndex]) {
+				return 1
+			}
+
+			return -1
+		}
+
+		if v.rightIndex >= len(v.right) {
+			// Ran out of the right array; if the left array has an identifier, left goes first, otherwise right
+			if unicode.IsLetter(v.left[v.leftIndex]) {
+				return -1
+			}
+
+			return 1
+		}
+
+		if unicode.IsDigit(v.left[v.leftIndex]) && unicode.IsDigit(v.right[v.rightIndex]) {
+			// Found the start of a number
+			compare := v.compareNumeric()
+			if compare != 0 {
+				return compare
+			}
+
+			continue
+		}
+
+		if unicode.IsLetter(v.left[v.leftIndex]) && unicode.IsLetter(v.right[v.rightIndex]) {
+			// Found the start of an identifier
+			compare := v.compareIdentifier()
+			if compare != 0 {
+				return compare
+			}
+
+			continue
+		}
+
+		if v.left[v.leftIndex] == v.right[v.rightIndex] {
+			// Both runes the same, skip to the next one
+			v.leftIndex++
+			v.rightIndex++
+			continue
+		}
+
+		// Runes are different, make a decision
+		if v.left[v.leftIndex] < v.right[v.rightIndex] {
```

Comment:
you reuse `v.left[v.leftIndex]` and `v.right[v.rightIndex]` quite a lot here. Wondering if it makes sense to (after your bounds checks) do:
```
leftRune := v.left[v.leftIndex]
rightRune := v.right[v.rightIndex]
```

It doesn't necessarily save you a lot of text but I think it saves a good amount of cognitive load when parsing the statements.